### ISSUE TITLE
bug: pipelines defined/registered with mismatched python versions (sometimes) fail mysteriously

### ISSUE THREAD

**Issue by OwenPriceSkelly** [Garden-AI/garden#224]
Update: turns out the behavior here is not notebook-specific, and seems to be due to a "forwards incompatibility" between the python version registering the pipeline and the version in the container (e.g. registering with 3.10.10 and running 3.9.16 in the container here). I think this is likely not a problem if the registering version is less than or equal to than the container version. 

acceptance criteria: 
- test this more exhaustively to see if we only need to worry about "forwards" incompatibilities, or if this can be reproduced with the registration/container versions mismatched in the other direction too 
- in cases we can know that this will happen (so far at least 3.10.10 -> 3.9.16 seems always problematic, even for dummy pipelines) the warning should be upgraded to an error
- include this in some troubleshooting docs

original issue for posterity:

tl;dr something about jupyter/dill interactions at a low level breaks remote pipeline execution in mysterious ways if the pipeline was defined and registered in a notebook, including a rare pure-python segfault when deserializing locally and an implausible `AttributeError: 'dict_itemiterator' object has no attribute 'environ'` when deserializing remotely.

When trying to execute a pipeline which was registered directly from a notebook, like this example:

<img width="970" alt="Screenshot 2023-07-18 at 12 05 10 PM" src="https://github.com/Garden-AI/garden/assets/21372141/a8d79bfc-7834-4ae6-91c5-cc4c5d13a80d">

we get the mysterious error message 
```
TaskExecutionFailed:
    Traceback (most recent call last):
      File "<string>", line 35, in inner
    AttributeError: 'dict_itemiterator' object has no attribute 'environ'
```

The `File "<string>"` bit is because the error is coming from something we exec'd in `__main__`, which seems to be the `garden_ai_compose_functions` helper (which does indeed access `os.environ` on its 35th line, but `os` is a builtin module not a `dict_itemiterator`, so ???). 

Adding to the mystery: an identical pipeline, registered from a pipeline.py (literally copy-pasted, works both when registered via CLI or client) runs fine on the same endpoint (dlhub) in the same fixed container ("7172db12-791c-4e1d-b675-07218f89c82b", has torch 1.8.0 and numpy 1.21.1 pinned). 

<img width="748" alt="Screenshot 2023-07-18 at 12 31 14 PM" src="https://github.com/Garden-AI/garden/assets/21372141/6276feb6-2deb-4c44-b10e-fd63e4e5ad4e">

**Comment by OwenPriceSkelly** [comment]
update: almost certainly something with the notebook is weird, when I try and register like this (from the notebook):
```python
bragg_ok = garden_ai.utils._meta._load_pipeline_from_python_file("pipeline.py")
client.register_pipeline(bragg_ok, container_id)
```
I reproduce the error that I couldn't otherwise reproduce with the `pipeline.py` version

**Comment by OwenPriceSkelly** [comment]
update: I can consistently reproduce this by registering the following extremely simple pipeline from both a regular script and from a jupyter notebook: 

```
@step
def run_inference(
    input_arg: object,
) -> object:
    return input_arg


dummy_pipeline: Pipeline = client.create_pipeline(
    title="dummy pipeline",
    steps=(run_inference,),
    authors=["Owen"],
    doi="10.23677/jcc7-5r19"
    short_name="dummy_pipeline",
)
```

for debugging, I tweaked globus compute to additionally dump the dill-serialized function to a file in both cases. Not only does the version serialized from a notebook fail when called, it even **segfaults**: 
```
  with open("direct_registration.dill", "rb") as fp:
      content = fp.read()
      func = dill.loads(content)
      print(func(42))  # prints 42

  with open("notebook_registration.dill", "rb") as fp:
      content = fp.read()
      func = dill.loads(content)
      print(func(42))  # Job 1, 'python -m pipeline' terminated by signal SIGSEGV (Address boundary error)
```

**Comment by OwenPriceSkelly** [comment]
update: silly not to try this sooner, but turns out that deserializing from a notebook instead of a script has a problem with the opposite serialization environment -- the `notebook_registration.dill` deserializes and executes fine in a notebook. Since that won't be the case on the remote endpoint, I think we may just not be able to support pipelines defined in notebooks. Maybe in the future we can do something clever to figure out what kind of environment we're running in, and write the user's pipeline to a tempfile or something so we can register it with a separate non-notebook process?

**Comment by OwenPriceSkelly** [comment]
another update: apparently this isn't a universal problem -- @isaac-darling hasn't had any trouble registering/executing his `foundry_example` seedling, which breaks for me (in basically same ways as the other pipelines I've tried above)

**Comment by WillEngler** [comment]
Closing as stale - I don't think this Python mismatch is a possibility any more
